### PR TITLE
fix: start height of initial public randomness commit

### PR DESCRIPTION
**Pull request by bap2pecs**
## Summary

see discussion here https://github.com/babylonchain/finality-provider/pull/447#discussion_r1660074723

let's first list the facts and requirements:

facts:
- first PR submissoin is based on tip height (e.g. height 27)
- chain poller starts from the lowest non-finalized height that the fp has not processed. (e.g. height 2)
- finalitySigSubmissionLoop receives signal from chain poller to try to submit sig from height 2 which will ofc fatal on max retry

requirements:
- we want to ensure no blocks with voting power will be missed

---

so I think the root cause is first PR submission shouldn't be determined by tipHeight. b/c it means the PR and Sig submission loops is not consistent in the start state.

my proposed fix is to add a global state firstPolledHeight that only be set once on the first poll to trigger the first PR submission

## Test Plan

comment out our changes in QueryFinalityProviderVotingPower first

```
make test-e2e-op
```

see test passed
```
$ make test-e2e-op                                                                    [15:22:07]
No processes to kill
rm -rf ~/.babylond ~/.wasmd ~/.bcd
cd tools; \
	go install -trimpath github.com/babylonchain/babylon/cmd/babylond
go test -mod=readonly -timeout=25m -v github.com/babylonchain/finality-provider/itest/opstackl2 -count=1 --tags=e2e_op --run ^TestBlockBabylonFinalized
=== RUN   TestBlockBabylonFinalized
    babylon_node_handler.go:148: babylon log file: /var/folders/9_/q4wsdnh14_s60_74cd2rbztm0000gp/T/zBabylonTest1670807382/babylon.log
    op_test_manager.go:83: Fp home dir: /var/folders/9_/q4wsdnh14_s60_74cd2rbztm0000gp/T/fpe2etest1412267153/fp-home
    setup.go:490: Generating L2 genesis l2_allocs_mode delta
    op_test_manager.go:109: Register consumer op-stack-l2-901 to Babylon
    op_test_manager.go:153: Deployed op finality contract address: bbn14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9sw76fy2
    op_test_manager.go:241: Babylon node has started
    op_test_manager.go:339: Registered Finality Provider b3996108efd9c82ca02dcc521fb39e20c2278ebf9ab58e9382b7ed962df9f64f for chain-test
    op_test_manager.go:383: The test manager is running with 1 finality-provider(s)
    op_test_manager.go:339: Registered Finality Provider 1c4b9c180ad2e9aa832e593490fa1d19e0b5fc7e81a68f12cd30efdcf12e7990 for op-stack-l2-901
    op_test_manager.go:339: Registered Finality Provider 08f021d4433b713c638fbe150f3446932ef01263de130db32ade5815b18eaba2 for op-stack-l2-901
    op_test_manager.go:383: The test manager is running with 3 finality-provider(s)
    utils.go:101: Public randomness for fp 08f021d4433b713c638fbe150f3446932ef01263de130db32ade5815b18eaba2 is successfully committed at height 65
    utils.go:101: Public randomness for fp 08f021d4433b713c638fbe150f3446932ef01263de130db32ade5815b18eaba2 is successfully committed at height 129
    utils.go:101: Public randomness for fp 1c4b9c180ad2e9aa832e593490fa1d19e0b5fc7e81a68f12cd30efdcf12e7990 is successfully committed at height 193
    utils.go:101: Public randomness for fp 1c4b9c180ad2e9aa832e593490fa1d19e0b5fc7e81a68f12cd30efdcf12e7990 is successfully committed at height 257
    base_test_manager.go:174: successfully submitted a BTC delegation
    base_test_manager.go:174: successfully submitted a BTC delegation
    base_test_manager.go:207: delegations are pending
    base_test_manager.go:227: delegations are active
    op_test_manager.go:284: Test block height 706 and 706
    op_test_manager.go:314: Submit finality signature to op finality contract &{Height:706 Hash:[163 198 161 192 245 28 28 32 201 102 21 235 1 197 250 177 144 130 99 49 168 224 213 83 0 193 77 184 49 152 223 82]}
    op_test_manager.go:314: Submit finality signature to op finality contract &{Height:706 Hash:[163 198 161 192 245 28 28 32 201 102 21 235 1 197 250 177 144 130 99 49 168 224 213 83 0 193 77 184 49 152 223 82]}
    op_e2e_test.go:190: Test case 1: block 706 is finalized
    op_test_manager.go:314: Submit finality signature to op finality contract &{Height:707 Hash:[185 213 238 184 77 103 140 11 128 213 132 155 49 115 2 116 83 205 111 62 204 148 199 207 229 103 136 55 144 13 147 230]}
    op_e2e_test.go:210: Test case 2: block 707 is not finalized
    op_test_manager.go:464: Stopping test manager
--- PASS: TestBlockBabylonFinalized (254.95s)
PASS
ok  	github.com/babylonchain/finality-provider/itest/opstackl2	256.528s
```

### CODE REVIEW

**Comment by bap2pecs** [comment]
will fix test after getting some feedback on this approach

**Comment by bap2pecs** [comment]
just had a discussion w @gitferry offline. the problem of this issue is there is a subtle assumption: when a FP first started, it doesn't have any voting power yet.

given that when we add timestamping later, the entire logic here for handling FP startup will need a massive refactoring, it's not worth it to make this temporary fix.

for now, we can just modify our e2e tests to satisfy this assumption

**Comment by bap2pecs** [comment]
@gitferry and I discussed again and decided to fix it before the timestamping feature is added.

File: finality-provider/service/fp_instance.go [hunk 1]
```diff
@@ -283,21 +283,50 @@ func (fp *FinalityProviderInstance) finalitySigSubmissionLoop() {
...
+			return
+		}
+		if txRes == nil {
+			fp.logger.Fatal(
+				"Error submitting the first randomness",
+				zap.String("consumer_id", string(fp.GetChainID())),
+				zap.String("pk", fp.GetBtcPkHex()),
+			)
+			return
+		}
```
Comment (gitferry) [inline finality-provider/service/fp_instance.go:313]: Is this check necessary? This case can happen when there's already sufficient randomness. Besides, `txRes` is not used in other places

File: finality-provider/service/fp_instance.go [hunk 2]
```diff
@@ -883,17 +956,14 @@ func (fp *FinalityProviderInstance) getPollerStartingHeight() (uint64, error) {
 	if err != nil {
```
Comment (gitferry) [inline finality-provider/service/fp_instance.go:956]: the doc string above can be modified

**Review by gitferry — APPROVED** [review]
Great fix! Some nits:

File: finality-provider/service/fp_instance.go (hunk 1 above)
Comment (bap2pecs) [reply finality-provider/service/fp_instance.go:313]: > This case can happen when there's already sufficient randomness

that's for the other method i.e. `retrySubmitFinalitySignatureUntilBlockFinalized`

in this one, it should never happen b/c we checked `lastCommittedHeight == uint64(0)` above. but I still want to have this check to be safe b/c `retryCommitPubRandUntilMaxRetry` can return `nil`

this is due to `CommitPubRand` can return `nil`. but actually I think it's better to add the check inside `retryCommitPubRandUntilMaxRetry`. let me fix it

> Besides, txRes is not used in other places

good catch. let me remove it

Comment (bap2pecs) [reply finality-provider/service/fp_instance.go:313]: actually, it doesn't make sense to add the check inside `retryCommitPubRandUntilMaxRetry` b/c we shouldn't assume the input argument is always for the case when there is no PR committed yet.

there are only two cases where both `txRes` and `err` will be `nil`

1) when there is sufficient PR. not that's not right b/c when this code is hit, there is no PR at all. so something very bad must have happened
2) when `CommitPubRandList()` returns `nil, nil`. this is not right b/c if there is no error, we should expect the `TxResponse` not to be `nil`

so if we need to check `txRes`, we cannot remove it

**Comment by bap2pecs** [comment]
```
fp_instance.go:771: Unexpected call to *mocks.MockConsumerController.CommitPubRandList([0xc0000853b0 26 100 [47 103 60 142 28 203 22 171 80 137 209 89 214 181 20 214 208 28 246 52 54 180 112 52 24 68 224 145 195 219 12 131] 0xc0022aa9b0]) at /home/circleci/project/finality-provider/service/fp_instance.go:771 because: there are no expected calls of the method "CommitPubRandList" for that receiver
    --- FAIL: FuzzRegisterFinalityProvider/seed#7 (0.36s)
```

this failure seems relevant

File: finality-provider/service/fp_instance.go (hunk 1 above)
Comment (gitferry) [reply finality-provider/service/fp_instance.go:313]: OK then. Let's keep the check

**Comment by bap2pecs** [comment]
ok just fixed a bunch of tests

**Comment by gitferry** [comment]
Thanks for fixing the tests. A kind reminder is to better merge it to `dev` (title could be reworded)

**Comment by bap2pecs** [comment]
@gitferry sure can I cherry-pick it to `dev` later in a separate PR?

**Comment by gitferry** [comment]
> any suggestions?

Something like `fix: start height of initial public randomness commit`?

**Comment by bap2pecs** [comment]
> > any suggestions?
> 
> Something like `fix: start height of initial public randomness commit`?

ah I see you point. the title was indeed inappropriate since now we don't add the extra internal state in the poller

File: finality-provider/service/fp_instance.go (hunk 1 above)
Comment (bap2pecs) [reply finality-provider/service/fp_instance.go:313]: jfyi this already detects the same issue we found here https://github.com/babylonchain/finality-provider/issues/450

```
    op_test_manager.go:285: Submit finality signature to op finality contract &{Height:642 Hash:[22 13 54 122 191 151 237 207 57 27 24 204 19 129 71 148 217 144 162 163 102 162 194 9 60 109 71 72 12 74 216 164]}
    op_e2e_test.go:210: Test case 2: block 642 is not finalized
    op_test_manager.go:435: Stopping test manager
2024-07-02T13:26:50.818Z        FATAL   service/fp_instance.go:307      Error submitting the first randomness   {"consumer_id": "chain-test", "pk": "b4209be0ad8cc86c658d3e6ff0971e90926c5869dc36494b5e8066f186434f22"}
github.com/babylonchain/finality-provider/finality-provider/service.(*FinalityProviderInstance).randomnessCommitmentLoop
        /home/circleci/project/finality-provider/service/fp_instance.go:307
FAIL    github.com/babylonchain/finality-provider/itest/opstackl2       337.629s
```

**Pull request merged**